helm-v3/006-repositories.md

```diff
@@ -1,40 +1,177 @@
+
+This will allow Helm to support any number of things in the future related
+to repositories without making breaking changes (e.g. remote `helm search`).
+
+### `/.well-known/helm`
+
+The Capabilities API will be enabled for a specific domain via
+[Well-Known URI](https://tools.ietf.org/html/rfc5785) located at
+`/.well-known/helm`.
+
+*Note: This file may also serve to provide configuration for other APIs
+in the future by supplying an additional top-level section.*
+
+The response body should be valid JSON and should supply a valid,
+top-level `capabilities` section. For example:
+
+```
+{
+    "capabilities": {
+        "login": {
+            ...
+        },
+        "push": {
+            ...
+        },
+        "search": {
+            ...
+        }
+    }
+}
+```
+
+Each subsection (i.e. `login`, `push`, `search`) will provide configuration
+specific to the capabiltity.
+
+For example, the following might inform Helm that this domain only
+accepts token-based login:
+
+```
+...
+        "login": {
+            "methods": {
+                "token": {
+                    "loginUrl": "https://auth.site.com/oauth2/authorize",
+                    "tokenUrl": "https://auth.site.com/oauth2/token"
+                }
+            }
+        }
+...
+```
+
+## The `helm login` command
+
+A new `helm login` command will be added to Helm, responsible for
+managing authentication across repository providers.
+
+The basic usage will be:
+
+```
+helm login <url>
+```
+
+This command is intended to allow users to authenticate just one time in
+order to access multiple repositories provided by the same source. 
+
+Afterwards, running a `helm repo add` pointing to any repository URL
+tied to a provider that has been successfully authenticated against will
+not require additional credentials.
+
+The exact semantics are TBD, but several auth methods will be supported,
+such as:
+
+* Basic Auth
+* Cert based Auth
+* Token/Bearer Auth
+
+### Storing Credentials
+
+Credentials obtained via `helm login` will be stored in a new `providers`
+section in `$HELM_HOME/repository/repositories.yaml`.
+
+For all provider entries, the following field is required:
+* `domain` - the FQDN of the original URL specified during `helm login`
+
+Other fields will be stored based on the authentication method used.
+
+The following example shows a snippet of  what `repositories.yaml` might
+look like with a `providers` section:
+```
+apiVersion: v1
+generated: 2018-10-25T17:02:41.478448894-05:00
+providers:
+- domain: site.com
+  realm: https://site.com/oauth2/helm
+  expires: 2018-10-26T17:02:41.478448894-05:00
+  accessToken: MTQ0NjJkZmQ5OTM2NDE1ZTZjNGZmZjI3
+  refreshToken: IwOGYzYTlmM2YxOTQ5MGE3YmNmMDFkNTVk
+repositories:
+...
+```
+
+## The `helm push` command
+
+A new `helm push` command will be added to Helm, responsible for uploading
+chart versions to a repository.
+
+The basic usage will be:
+
+```
+helm push <chart> <repo>
```

Comment:
I'm assuming this looks like `helm push helloworld demo42` assuming a did a `helm repo add of demo42.azurecr.io`
Since this is a new API, and I was hoping to avoid the local store, could this be:
`helm push demo42.azurecr.io/project1/subproject/helloworld:1.0.0`

Reply:
On app-registry we tried`helm push <chart> <repo>` but finally did: `helm push $registry/$namespace`
like:
```
helm registry push quay.io/ant31/ 
helm registry push localhost:5000/localrepo
```

And to pull also the full path:
```
helm registry push quay.io/ant31/jenkins@v3.4.5 
```

Reply:
I think if we plan to support existing UX, we can make the `<repo>` positional arg optional. Then if the `<chart>` arg looks like a URL as you've described, we treat it as such.

`helm fetch` already supports pull from URL, push should allow for the same experience.